cctp_primitives/src/bit_vector/merkle_tree.rs

```diff
@@ -0,0 +1,163 @@
+//! # Merkle Tree
+//!
+//! `merkle_tree` exposes functions to compute a bit vector Merkle tree.
+
+use super::compression;
+
+use algebra::{
+    fields::tweedle::Fr as TweedleFr, ToConstraintField
+};
+
+use primitives::{
+    crh::poseidon::parameters::tweedle::{TweedleFrPoseidonHash, TweedleFrBatchPoseidonHash},
+    merkle_tree::field_based_mht::{
+        FieldBasedMerkleTreeParameters, FieldBasedMerkleTreePrecomputedEmptyConstants,
+        BatchFieldBasedMerkleTreeParameters, FieldBasedOptimizedMHT, FieldBasedMerkleTree,
+        parameters::tweedle_fr::TWEEDLE_MHT_POSEIDON_PARAMETERS
+    }
+};
+
+use bit_vec::BitVec;
+
+#[derive(Clone, Debug)]
+struct TweedleFieldBasedMerkleTreeParams;
+impl FieldBasedMerkleTreeParameters for TweedleFieldBasedMerkleTreeParams {
+    type Data = TweedleFr;
+    type H = TweedleFrPoseidonHash;
+    const MERKLE_ARITY: usize = 2;
+    const EMPTY_HASH_CST: Option<FieldBasedMerkleTreePrecomputedEmptyConstants<'static, Self::H>> = Some(TWEEDLE_MHT_POSEIDON_PARAMETERS);
+}
+
+impl BatchFieldBasedMerkleTreeParameters for TweedleFieldBasedMerkleTreeParams {
+    type BH = TweedleFrBatchPoseidonHash;
+}
+
+type TweedlePoseidonMHT = FieldBasedOptimizedMHT<TweedleFieldBasedMerkleTreeParams>;
+
+type Error = Box<dyn std::error::Error>;
+
+const MERKLE_TREE_HEIGHT: usize = 12;
+
+/// Computes the root hash of the Merkle tree created as a representation
+/// of `uncompressed_bit_vector`.
+///
+/// The input byte array is expected to have a Big Endian bit order, for example:
+/// Bit Array [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0] <=> Byte Array [1, 2]
+/// 
+/// # Examples
+///
+/// ```
+/// use cctp_primitives::bit_vector::merkle_tree::*;
+///
+/// let bit_vector: Vec<u8> = (0..100).collect();
+/// let merkle_root = merkle_root_from_bytes(&bit_vector).unwrap();
+/// 
+/// ```
+pub fn merkle_root_from_bytes(uncompressed_bit_vector: &[u8]) -> Result<algebra::Fp256<algebra::fields::tweedle::FrParameters>, Error> {
+
+    let bv = BitVec::from_bytes(&uncompressed_bit_vector);
+    let bool_vector: Vec<bool> = bv.into_iter().map(|x| x).collect();
+
+    let num_leaves = 1 << MERKLE_TREE_HEIGHT;
+    let mut mt = TweedlePoseidonMHT::init(
+        MERKLE_TREE_HEIGHT,
+        num_leaves,
+    );
+
+    let leaves = bool_vector.to_field_elements()?;
```

Comment:
Two comments here:
1) When deserializing from bits, the bit array should have a big endian bit order;
2) Let's assume that who serializes the bitvector, serializes a leaf in CAPACITY bits, concatenates all the bits in a single bit array, and then pad with 0s to reach a length which is multiple of 8.
The to_constraint_field() function will read a FieldElement out of the padding bits (because it uses chunk() function internally), that is not what we want. So either you trim here the padding bits or you copy the to_constraint_field() implementation from GingerLib and you replace the chunks() function with chunk_exact().

Anyhow, I suggest to sync with @nastenko on both points and possibily let him create some test data that you can use here as a regression test, to be sure that the two implementations are compatible.

Reply:
Yes, the bits order of a serialized bitvector and the way of serialization is exactly as described in 1 and 2. The actual implementation is in `serialize_bvt` and `initialize_bvt` methods here:
https://github.com/HorizenOfficial/zendoo-sc-cryptolib/pull/43/files#diff-e1af7a26102f1419b80eae8645bbf1cfa603aa98cda6c1556a6cd4e778aa4f61

Reply:
Are you sure about it ? Because you use GingerLib's write_bits() function and then you reverse the bit array saying that you convert it to BigEndian form; however, write_bits() function already outputs the bit array in BigEndian form, so by reversing it you're actually converting it into LittleEndian

Reply:
Fixed.

After the latest changes made:
- Potential padding bits are removed from the bit vector before computing the merkle root;
- Added a check to verify that the bit vector size exactly fits the number of leaves in a merkle tree (i.e. the value `Bit_Vector_Size_Withouth_Padding / FIELD_CAPACITY` must be a power of 2;
- The unit test for the merkle root computation are performed by using known bit vectors (i.e. we are checking that we calculate the same root hash both on the CCTP Lib and on the SC Crypto Lib).